src/dbnode/storage/index/types.go

```diff
@@ -803,11 +841,19 @@ func (b *WriteBatch) MarkUnmarkedEntriesSuccess() {
 // MarkEntrySuccess marks an entry as success.
 func (b *WriteBatch) MarkEntrySuccess(idx int) {
 	if !b.entries[idx].result.Done {
+		indexedEntry, closer, reconciled := b.entries[idx].OnIndexSeries.ReconciledOnIndexSeries()
```

Comment:
The only worry here is the added shard map lock contention that this reconciliation costs. Before we land this we can test in some env at scale and get a sense of any perf impact.

Reply:
There's also the possibility that this is the entry that will be orphaned and that there is nothing in the shard map at all right now (in fact this is more likely since orphans are typically a side effect of a first insertion which doesn't win the race to be inserted into the shard map).

Instead of adding potential performance overhead to the write path, couldn't we do this same merge of states logic when we consider it for garbage collection and/or read reconciliation time?
https://github.com/m3db/m3/blob/82aa22e67816fe2dcf08c50c6380bfac8ce9a450/src/dbnode/storage/index/mutable_segments.go#L233-L247

I assume that's why we're doing this at all for higher accuracy in both read reconcilation and garbage collection, so wouldn't it be better to pay the cost in those parts of the code base?

Most things are architected so that write path stays lean which causes complexity downstream, but this seems to go against that pattern somewhat so I just want to understand if we can achieve the same thing but in code paths where we're willing to pay more cycles/latency (which for the background garbage collection path of the index we have plenty more cycles/latency to spare).